documentation/CAMARA API Design Guide.md

```diff
@@ -0,0 +1,1120 @@
+   TLS ensures the confidentiality of the transported data and that the server's hostname matches the server's SSL certificate.
+   - If HTTP 2 is used to improve performance, you can even send multiple requests over a single connection, this way you will avoid the complete overhead of TCP and SSL on later requests.
+
+3. **Using hash password**. 
+Passwords should never be sent in API bodies,
+   but if it is necessary it must be hashed to protect the system
+   (or minimize damage) even if it is compromised in some hacking attempts.
+   There are many hashing algorithms that can be really effective for password security,
+   for example, PBKDF2, bcrypt and scrypt algorithms.
+
+4. **Information must not be exposed in the URLs**
+Usernames, passwords, session tokens, and API keys should not appear in the URL, as this can be captured in web server logs, making them easily exploitable. For example, this URL (```https://api.domain.com/user-management/users/{id}/someAction?apiKey=abcd123456789```) exposes the API key. Therefore, never use this kind of security.
+
+5. **Authentication and authorization must be considered**
+   CAMARA uses the authentication and authorization protocols and flows as described in the [Camara Security and Interoperability Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md).
+
+6. **Add request time flags should be considered**. 
+Along with other request parameters, a request timestamp can be added as a custom HTTP header in API requests.
+   The server will compare the current timestamp with the timestamp of the request
+   and will only accept the request if it is within a reasonable time frame
+   (1–2 minutes maybe).
+   - This will prevent very basic replay attacks from people trying to hack your system without changing this timestamp.
+
+7. **Entry params validation**
+ Validates the parameters of the request in the first step before it reaches the application logic.
+   Put strong validation checks and reject the request immediately if validation fails.
+   In the API response,
+   send relevant error messages and an example of the correct input format to improve the user experience.
+
+### Security definition
+
+In general, all APIs must be secured to ensure who has access to what and for what purpose.
+CAMARA uses OIDC and CIBA for authentication and consent collection and to determine whether the user has,
+e.g. opted out of some API access.
+
+The [Camara Security and Interoperability Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#purpose) defines that a single purpose is encoded in the list of scope values. The purpose values are defined by W3C Data Privacy Vocabulary as indicated in the [Profile](https://github.com/camaraproject/IdentityAndConsentManagement/blob/main/documentation/CAMARA-Security-Interoperability.md#purpose-as-a-scope).

```

Comment:
Always? 
What about APIs, where OAuth Client Credentials (i.e. no relation to personal data) can be used?